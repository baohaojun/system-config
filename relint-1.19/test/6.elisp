;;; Relint test file 6          -*- emacs-lisp -*-

;; Test mixup of regexp and character alternative
(defun myrefun (things)
  (if (listp things)
      (regexp-opt things)
    things))

(defun test-mixup (x y z)
  (list
   (concat "a*[^" some-regexp "]")
   (concat stuff "[" (regexp-quote x) "]" stuff)
   (concat "[" (myrefun y) "]")
   (concat "[" (rx ?a "b" nonl) "]")
   (concat "[" (rx-to-string z) "]")
   (format "^%s[%s]*%s$" stuff some-regex stuff)
   (format "[%s]" (regexp-quote x))
   (format "[%s]" (myrefun y))
   (format "[%s]" (rx ?a "b" nonl))
   (format "[%s]" (rx-to-string z))
   (format "[%s]" (f (string-match (regexp-quote x) a))) ; ok
   (format "[%s]" (f (replace-regexp-in-string (regexp-quote x) a b))) ; ok
   (format "[%s]" (f (replace-regexp-in-string a b (regexp-quote x)))) ; bad
   ))

;; Test skip-chars
(defun test-skip-chars (x y z)
  (skip-chars-forward "[a-z]")
  (skip-chars-backward "aa")
  (skip-chars-forward some-re)
  (skip-chars-backward (regexp-quote x))
  (skip-chars-forward (myrefun y))
  (skip-chars-backward (rx "abc"))
  (skip-chars-forward (rx-to-string z)))

;; Test skip-syntax
(defun test-skip-syntax (x y z)
  (skip-syntax-forward "\\s-")
  (skip-syntax-backward "- ")
  (skip-syntax-forward some-re)
  (skip-syntax-backward (regexp-quote x))
  (skip-syntax-forward (myrefun y))
  (skip-syntax-backward (rx "w-"))
  (skip-syntax-forward (rx-to-string z)))

;; Test incorrect provenance tracing
(defun test-regexp-in-condition (x y z)
  (format "[%s]" (if (regexp-quote x) y z))
  (skip-chars-forward (when some-regexp y))
  (skip-syntax-backward (unless (myrefun x) y))
  (skip-chars-backward (while (myrefun x) z)))

;; Test suppression
(defun test-suppression ()
  ;; relint suppression: Unescaped literal .\$
  ;; relint suppression: Duplicated .a
  (looking-at "$[aa]"))

;; Test user-defined regexp-generating functions
(defun make-a-nice-regexp ()
  (stuff))

(defun make-something ()
  "Return a regexp made from whole cloth."
  (stuff))

(defun test-user-defined-generator ()
  (skip-chars-forward (make-a-nice-regexp))
  (skip-chars-backward (make-something)))
