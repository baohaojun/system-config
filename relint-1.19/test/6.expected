6.elisp:11:19: Value from `some-regexp' cannot be spliced into `[...]'
6.elisp:12:22: Value from `regexp-quote' cannot be spliced into `[...]'
6.elisp:13:16: Value from `regexp-opt' cannot be spliced into `[...]'
6.elisp:14:16: Value from `rx' cannot be spliced into `[...]'
6.elisp:15:16: Value from `rx-to-string' cannot be spliced into `[...]'
6.elisp:16:32: Value from `some-regex' cannot be spliced into `[...]'
6.elisp:17:19: Value from `regexp-quote' cannot be spliced into `[...]'
6.elisp:18:19: Value from `regexp-opt' cannot be spliced into `[...]'
6.elisp:19:19: Value from `rx' cannot be spliced into `[...]'
6.elisp:20:19: Value from `rx-to-string' cannot be spliced into `[...]'
6.elisp:23:19: Value from `regexp-quote' cannot be spliced into `[...]'
6.elisp:28:24: In call to skip-chars-forward: Suspect skip set framed in `[...]' (pos 0)
  "[a-z]"
   ^
6.elisp:29:26: In call to skip-chars-backward: Duplicated character `a' (pos 1)
  "aa"
   .^
6.elisp:30:23: `some-re' cannot be used for arguments to `skip-chars-forward'
6.elisp:31:24: `regexp-quote' cannot be used for arguments to `skip-chars-backward'
6.elisp:32:23: `regexp-opt' cannot be used for arguments to `skip-chars-forward'
6.elisp:33:24: `rx' cannot be used for arguments to `skip-chars-backward'
6.elisp:34:23: `rx-to-string' cannot be used for arguments to `skip-chars-forward'
6.elisp:38:27: In call to skip-syntax-forward: Invalid char `s' in syntax string (pos 1)
  "\\s-"
   ..^
6.elisp:39:27: In call to skip-syntax-backward: Duplicated syntax code ` ' (pos 1)
  "- "
   .^
6.elisp:40:24: `some-re' cannot be used for arguments to `skip-syntax-forward'
6.elisp:41:25: `regexp-quote' cannot be used for arguments to `skip-syntax-backward'
6.elisp:42:24: `regexp-opt' cannot be used for arguments to `skip-syntax-forward'
6.elisp:43:25: `rx' cannot be used for arguments to `skip-syntax-backward'
6.elisp:44:24: `rx-to-string' cannot be used for arguments to `skip-syntax-forward'
6.elisp:68:23: `make-a-nice-regexp' cannot be used for arguments to `skip-chars-forward'
6.elisp:69:24: `make-something' cannot be used for arguments to `skip-chars-backward'
